### ISSUE TITLE
Pre-Fetch Future Episodes for Currently Watched Show(s)

### ISSUE THREAD

**Issue by kevincosmos** [ApolloTVofficial/kamino#126]
I am proposing a new feature that would work in the following way:

User plays S1E1 of "TV Show A" in Apollo, and once they start playing E1, the server will, in the background, start searching links for episodes 2, 3, 4, 5 (or whatever is determined by the user in Settings) and caches those results in the DB for the user.

Once the user finishes E1 and goes to scrape for E2, the server has already done the work and will provide the results much quicker than otherwise would be possible. Once E2 plays, episode 6 is also scraped. So on and so forth.

Any content that is not IP locked can be done by default across the entire population by default, but for anything IP locked this feature would probably be disabled by default as users may have concerns about the app doing things for them in the background that they haven't necessarily given permission for. 

The user will be able to toggle turning this feature on, and toggle how many episodes in advance episodes can be scraped (within reason).

The server will only keep the links valid for "Y" amount of time determined by the devs.

**Comment by edmundackah** [comment]
That is potentially wasteful when you think of the exponential increase in API calls and how many people will actually go ahead to watch the following episodes. If implemented the user should toggle it from the settings.

**Comment by kevincosmos** [comment]
> That is potentially wasteful when you think of the exponential increase in API calls and how many people will actually go ahead to watch the following episodes. If implemented the user should toggle it from the settings.

So maybe we disable it for everyone by default?
If it's gonna be too much of a strain on the load/server, maybe only allow it for those who choose to go client-side rather than server-side?

**Comment by SpajicM** [comment]
How about limiting it to just the next episode? Instead of having a setting that you can do 5 episodes in advance, just do 'while watching E1, scrape E2,while watching E2, scrape E3'.

Same UX improvement, lot less useless calls. I'm definitely for implementation of this.

**Comment by ddivad** [comment]
I don't think it would be overly wasteful, but agree that 1 - 2 episodes in advance would be better than 4 - 5 in advance.

Also, I think we should disable by default and add the option to enable it, that way only users that want this functionality would create extra requests, not all users. I think this would reduce the server load, which shouldn't be very high anyway, as its just like a user doing 2 searches anyway. 

If the user does indeed go on to watch the next episode, we are saving an api call in that case, so really the load should balance out in most cases.

**Comment by Chaphasilor** [comment]
This would be also very useful if a 'continue watching'/'next episode' button ever were to be implemented.  
Just hit the button and the episode could start playing instantly, if combined with the autoplay feature

**Comment by NBTX** [comment]
What about 5 minutes before the end, or before the continue watching button is displayed, it starts scraping? That way we're more certain the user will want to watch the next episode.
